File: mini-coupang/backend/src/main/java/com/deepblog/minicoupang/domain/product/domain/Product.java
package com.deepblog.minicoupang.domain.product.domain;

import static jakarta.persistence.EnumType.STRING;
import static jakarta.persistence.FetchType.LAZY;
import static jakarta.persistence.GenerationType.IDENTITY;
import static lombok.AccessLevel.PRIVATE;
import static lombok.AccessLevel.PROTECTED;

import com.deepblog.minicoupang.domain.common.BaseEntity;
import com.deepblog.minicoupang.domain.product.exception.InvalidProductException;
import com.deepblog.minicoupang.domain.seller.domain.Seller;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import java.util.ArrayList;
import java.util.List;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Getter
@NoArgsConstructor(access = PROTECTED)
@AllArgsConstructor(access = PRIVATE)
@Builder
@Table(name = "products")
public class Product extends BaseEntity {

    @Id
    @GeneratedValue(strategy = IDENTITY)
    @Column(name = "product_id")
    private Long id;

    @ManyToOne(fetch = LAZY, optional = false)
    @JoinColumn(name = "seller_id", nullable = false)
    private Seller seller;

    @Column(name = "category_id", nullable = false)
    private Long categoryId;

    @Column(name = "name", nullable = false, length = 200)
    private String name;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "base_price", nullable = false)
    private Long basePrice;

    @Enumerated(STRING)
    @Column(name = "status", nullable = false, length = 20)
    private ProductStatus status;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<ProductOption> options = new ArrayList<>();

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<ProductImage> images = new ArrayList<>();

    public static Product create(
        Seller seller,
        Long categoryId,
        String name,
        String description,
        Long basePrice
    ) {
        validateSeller(seller);
        validateCategoryId(categoryId);
        validateName(name);
        validateBasePrice(basePrice);

        return Product.builder()
            .seller(seller)
            .categoryId(categoryId)
            .name(name)
            .description(description)
            .basePrice(basePrice)
            .status(ProductStatus.ACTIVE)
            .build();
    }

    private static void validateSeller(Seller seller) {
        if (seller == null) {
            throw new InvalidProductException("판매자 정보가 올바르지 않습니다.");
        }
    }

    private static void validateCategoryId(Long categoryId) {
        if (categoryId == null || categoryId <= 0) {
            throw new InvalidProductException("카테고리 정보가 올바르지 않습니다.");
        }
    }

    private static void validateName(String name) {
        if (name == null || name.isBlank()) {
            throw new InvalidProductException("상품명은 비어 있을 수 없습니다.");
        }
        if (name.length() < 2 || name.length() > 200) {
            throw new InvalidProductException("상품명은 2자 이상 200자 이하여야 합니다.");
        }
    }

    private static void validateBasePrice(Long basePrice) {
        if (basePrice == null) {
            throw new InvalidProductException("가격 정보가 필요합니다.");
        }
        if (basePrice < 0) {
            throw new InvalidProductException("가격은 0 이상이어야 합니다.");
        }
    }

    public void suspend() {
        if (this.status != ProductStatus.ACTIVE) {
            throw new InvalidProductException("판매 중인 상품만 정지할 수 있습니다.");
        }
        this.status = ProductStatus.SUSPENDED;
    }

    public void resume() {
        if (this.status != ProductStatus.SUSPENDED) {
            throw new InvalidProductException("정지된 상품만 판매를 재개할 수 있습니다.");
        }
        this.status = ProductStatus.ACTIVE;
    }

    public void markSoldOut() {
        if (this.status != ProductStatus.ACTIVE) {
            throw new InvalidProductException("판매 중인 상품만 품절 처리할 수 있습니다.");
        }
        this.status = ProductStatus.SOLD_OUT;
    }

    public ProductOption addOption(String optionName, String sku, Long additionalPrice) {
        assertSkuNotDuplicated(sku);
        ProductOption option = ProductOption.forProduct(this, optionName, sku, additionalPrice);
        this.options.add(option);
        return option;
    }

    private void assertSkuNotDuplicated(String sku) {
        if (sku == null) {
            return;
        }
        this.options.stream()
            .filter(existing -> sku.equals(existing.getSku()))
            .findAny()
            .ifPresent(existing -> {
                throw new InvalidProductException("이미 등록된 SKU입니다: " + sku);
            });
    }

    public ProductImage addImage(String url, boolean isPrimary) {
        if (isPrimary) {
            assertNoExistingPrimaryImage();
        }
        int nextOrdering = this.images.size();
        ProductImage image = ProductImage.forProduct(this, url, nextOrdering, isPrimary);
        this.images.add(image);
        return image;
    }

    private void assertNoExistingPrimaryImage() {
        this.images.stream()
            .filter(ProductImage::isPrimary)
            .findAny()
            .ifPresent(existing -> {
                throw new InvalidProductException("대표 이미지는 하나만 등록할 수 있습니다.");
            });
    }
}

File: services/product-service/src/main/kotlin/com/deepblog/product/domain/Product.kt
package com.deepblog.product.domain

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Index
import jakarta.persistence.Table
import jakarta.persistence.Version

/**
 * Product 도메인 엔티티.
 *
 * aggregate 설계 (`project-architecture.md` §2 + `domain-design.md` §4):
 * - Seller와는 다른 aggregate. `sellerId: Long`로 ID 참조만 보관.
 * - `sellerName: String`은 등록 시점 **스냅샷**. Seller가 나중에 이름을
 *   바꿔도 과거 상품에는 소급 적용하지 않음(조회 QPS + 판매자 재명명
 *   빈도 낮음 가정). 실 환경에서 판매자 변경 이벤트 도착 시 upsert 전략은
 *   Phase 2 이후에 도입.
 * - price는 원화 단위 `Long` (소수점 없음).
 * - @Version optimistic locking은 Phase 2 재고 차감 동시성 실험용 기반.
 */
@Entity
@Table(
    name = "products",
    indexes = [
        Index(name = "ix_products_seller_id", columnList = "seller_id"),
    ],
)
class Product(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    @Column(name = "seller_id", nullable = false)
    val sellerId: Long,

    @Column(name = "seller_name", nullable = false, length = 80)
    val sellerName: String,

    @Column(nullable = false, length = 200)
    var name: String,

    @Column(nullable = false)
    var price: Long,

    @Column(nullable = false)
    var stock: Int,

    @Version
    var version: Long = 0,
) {
    init {
        require(price >= 0) { "price must be non-negative: $price" }
        require(stock >= 0) { "stock must be non-negative: $stock" }
    }
}
